listenbrainz_spark/recommendations/recommend.py

```diff
@@ -0,0 +1,139 @@
+    user_id = get_user_id(user_name)
+    user_playcounts = run_query("""
+        SELECT recording_id
+          FROM playcount
+         WHERE user_id = %d
+    """ % user_id)
+
+    candidate_recordings = all_recordings.subtract(user_playcounts) 
+    candidate_recordings.limit(count)
+    candidate_recordings = candidate_recordings.rdd.map(lambda r : r['recording_id'])
+
+    recommendations = model.predictAll(candidate_recordings.map(lambda recording: (user_id, recording))).takeOrdered(50, lambda product: -product.rating)
+    recommended_recording_ids = [(recommendations[i].product) for i in range(len(recommendations))]
+
+    recommendations_df = recordings_df[recordings_df.recording_id.isin(recommended_recording_ids)].collect()
+    recommended_recordings = []
+
+    for row in recommendations_df:
+        rec = (row.track_name, row.recording_msid, row.artist_name, row.artist_msid, row.release_name, row.release_msid)
+        recommended_recordings.append(rec)
+
+    user_recommendations = defaultdict(dict)
+    user_recommendations['recordings'] = recommended_recordings
+    return user_recommendations
+
+def main():
+    ti = time.time()
+    try:
+        listenbrainz_spark.init_spark_session('Recommendations')
+    except Exception as err:
+        raise SystemExit("Cannot initialize Spark Session: %s. Aborting..." % (str(err)))
+    
+    try:
+        path = os.path.join('/', 'data', 'listenbrainz', 'recommendation-engine', 'dataframes')
+        playcounts_df = listenbrainz_spark.sql_context.read.parquet(config.HDFS_CLUSTER_URI + path + '/playcounts_df.parquet')
+        users_df = listenbrainz_spark.sql_context.read.parquet(config.HDFS_CLUSTER_URI + path + '/users_df.parquet')
+        recordings_df = listenbrainz_spark.sql_context.read.parquet(config.HDFS_CLUSTER_URI + path + '/recordings_df.parquet')
+    except Exception as err:
+        raise SystemExit("Cannot read dataframes from HDFS: %s. Aborting..." % (str(err)))
+
+    time_info = defaultdict(dict)
+    t = "%2.f" % ((time.time() - ti) / 60)
+    time_info['dataframes'] = t 
+    users_df.createOrReplaceTempView('user')
+    playcounts_df.createOrReplaceTempView('playcount')
+    recordings_df.persist()
+    count = recordings_df.count() // 2
+
+    t0 = time.time()
+    all_recordings = recordings_df.select('recording_id')
+    all_recordings.persist()
+    all_recordings_count = "{:,}".format(all_recordings.count())
+    t = "%.2f" % ((time.time() - t0) / 60)
+    time_info['all_recordings'] = t
+
+    path = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'recommendation-metadata.json')
+    with open(path, 'r') as f:
+        recommendation_metadata = json.load(f)
+        best_model_id = recommendation_metadata['best_model_id']
+
+    best_model_path = os.path.join('/', 'data', 'listenbrainz', 'recommendation-engine', 'best_model', '{}'.format(best_model_id))
+    
+    print("Loading model...")
+    for attempt in range(config.MAX_RETRIES):
```

Comment:
Gah, this is ugly. (not your fault). I wonder if this gets better when we run on azure.

Reply:
What can be an alternative to this?

Reply:
I suspect that this is a cause of poor performance of our VMs. An alternative is to rent better VMs.